src/benchmarks/micro/libraries/System.Runtime/Perf.Half.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+using BenchmarkDotNet.Attributes;
+
+using MicroBenchmarks;
+
+namespace System.Tests
+{
+    [BenchmarkCategory(Categories.Libraries)]
+    public class Perf_Half
+    {
+        private float[] bufferSourceSingle;
+        private float[] bufferDestinationSingle;
+        private Half[] bufferDestinationHalf;
+        private Half[] bufferSourceHalf;
+
+        public static IEnumerable<Half> Values => new Half[]
+        {
+            BitConverter.UInt16BitsToHalf(0x03ff),  //Maximum subnormal number in Half
+            (Half)12345.0f /* same value used by other tests to compare the perf */,
+            BitConverter.UInt16BitsToHalf(0x7dff)   //NaN
+        };
+
+        public static IEnumerable<float> SingleValues => new float[]
+        {
+            6.097555E-05f,
+            12345.0f /* same value used by other tests to compare the perf */,
+            65520.0f,   //Minimum value that is infinity in Half
+            float.NaN
+        };
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            const int Samples = 65536;
+            bufferDestinationSingle = new float[Samples];
+            Half[] bSH = bufferSourceHalf = new Half[Samples];
+            Span<Half> spanSH = bSH.AsSpan();
+            for (var i = 0; i < spanSH.Length; i++)
+            {
+                spanSH[i] = BitConverter.UInt16BitsToHalf((ushort)i);
+            }
+            ref Half x9 = ref MemoryMarshal.GetReference(spanSH);
+            int length = spanSH.Length;
+            int olen = length - 2;
+            var rng = new Random(12345);    //ValuesGenerator doesn't support exhaustive permutation
+            for (var i = 0; i < olen; i++)
+            {
+                int x = rng.Next(i, length);
+                (Unsafe.Add(ref x9, x), Unsafe.Add(ref x9, i)) = (Unsafe.Add(ref x9, i), Unsafe.Add(ref x9, x));
+            }
+            bufferDestinationHalf = new Half[Samples];
+            bufferSourceSingle = bSH.Select(a => (float)a).ToArray();
+        }
+        #region Permuted
+
+        [Benchmark]
+        public void SingleToHalfPermutedSimple()
+        {
+            Span<float> bA = bufferSourceSingle.AsSpan();
+            Span<Half> bD = bufferDestinationHalf.AsSpan();
+            ref float rsi = ref MemoryMarshal.GetReference(bA);
+            ref Half rdi = ref MemoryMarshal.GetReference(bD);
+            nint i = 0, length = Math.Min(bA.Length, bD.Length);
+            for (; i < length; i++)
+            {
+                Unsafe.Add(ref rdi, i) = (Half)Unsafe.Add(ref rsi, i);
+            }
+        }
```

Comment:
We shouldn't be looping or doing other complex logic here unless absolutely necessary. If customization is absolutely required then extensibility points such as `OperationsPerInvoke`, `IterationSetup`, and `IterationCleanup` should be preferred over more logic in the actual benchmarked method. -- That being said, this shouldn't be one of those cases, we really only hit such cases for intrinsics where we're testing a single instruction that takes 1-3 cycles to execute.

Doing "more" here means we aren't measuring the cost of `float->half` conversions, but rather we're measure the cost of `Math.Min`, `AsSpan`, `GetReference`, `looping`, and `float->half`.

We really instead want something like:
```csharp
[ParamsSource(nameof(HalfValues))]
public Half HalfValue;

public IEnumerable<Half> HalfValues => new Half[] { ... };

[Benchmark]
public float HalfToSingle(Half value) => (float)value;

[Params(...)]
public float SingleValue;

[Benchmark]
public Half SingleToHalf(float value) => (Half)value;
```

There are also options like `[Arguments]` and `[ArgumentsSource]` that can work as well: https://benchmarkdotnet.org/articles/features/parameterization.html

Reply:
CC. @adamsitnik

Reply:
The purpose of randomizing the order of the inputs was to reveal the dependence of performance on the order of the inputs.
In practical use, the same value is almost never given as an argument over and over again in succession, as in benchmarking. In most cases, a wide variety of data is given as arguments.
In such situations, branch predictions that rely on the consistency of input values should not be relied upon, and algorithms that use as few branches as possible should be favored in both reality and the benchmark.

Reply:
I created an Issue https://github.com/dotnet/BenchmarkDotNet/issues/2314 to replace loops by it.